tiledb/api/src/query/read/mod.rs

```diff
@@ -0,0 +1,361 @@
+    {
+        let scratch = scratch_allocator.alloc();
+
+        let scratch = OutputLocation {
+            data: BufferMut::Owned(scratch.0),
+            cell_offsets: scratch.1.map(BufferMut::Owned),
+        };
+
+        let scratch = Box::pin(RefCell::new(scratch));
+
+        let base = {
+            let scratch = scratch.as_ref().get_ref()
+                as *const RefCell<
+                    OutputLocation<'data, <T as ReadCallback>::Unit>,
+                >;
+            let scratch = unsafe {
+                &*scratch
+                    as &'data RefCell<
+                        OutputLocation<'data, <T as ReadCallback>::Unit>,
+                    >
+            };
+            self.register_callback((field.as_ref(), scratch), callback)
+        }?;
+
+        Ok(ManagedReadBuilder {
+            alloc: scratch_allocator,
+            scratch,
+            base,
+        })
+    }
+
+    /// Register a typed result to be constructed from the query results.
+    /// Intermediate raw results are written into the provided scratch space.
+    fn register_constructor<'data, S, T>(
+        self,
+        field: S,
+        scratch: &'data RefCell<
+            OutputLocation<
+                'data,
+                <<T as ReadResult>::Constructor as ReadCallback>::Unit,
+            >,
+        >,
+    ) -> TileDBResult<TypedReadBuilder<'data, T, Self>>
+    where
+        S: AsRef<str>,
+        T: ReadResult,
+        <T as ReadResult>::Constructor: Default,
+    {
+        let r = <T::Constructor as Default>::default();
+        Ok(TypedReadBuilder {
+            _marker: std::marker::PhantomData,
+            base: self.register_callback((field.as_ref(), scratch), r)?,
+        })
+    }
+
+    /// Register a typed result to be constructed from the query results.
+    /// Scratch space for raw results is managed by the callback.
+    fn register_constructor_managed<'data, S, T, R, C, A>(
+        self,
+        field: S,
+        scratch_allocator: A,
+    ) -> TileDBResult<
+        ManagedReadBuilder<'data, C, A, TypedReadBuilder<'data, T, Self>>,
+    >
+    where
+        S: AsRef<str>,
+        T: ReadResult<Constructor = R>
+            + HasScratchSpaceStrategy<C, Strategy = A>,
+        R: Default + ReadCallback<Unit = C>,
+        A: ScratchAllocator<C>,
+    {
+        let scratch = scratch_allocator.alloc();
+
+        let scratch = OutputLocation {
+            data: BufferMut::Owned(scratch.0),
+            cell_offsets: scratch.1.map(BufferMut::Owned),
+        };
+
+        let scratch = Box::pin(RefCell::new(scratch));
+
+        let base = {
+            let scratch = scratch.as_ref().get_ref()
+                as *const RefCell<OutputLocation<'data, C>>;
+            let scratch = unsafe {
+                &*scratch as &'data RefCell<OutputLocation<'data, C>>
+            };
+            self.register_constructor::<S, T>(field, scratch)
+        }?;
+
+        Ok(ManagedReadBuilder {
```

Comment:
They aren't quite the same, since we pass `self` to `register_[constructor|callback]` and get a different type back out from each of those, and that's what is passed to the managed reader base.